### PR TITLE
.Net: ADR for function call content model classes

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
Today, in SK, LLM function calling is supported exclusively by the OpenAI connector, and the function calling model is specific to that connector. The new AI connectors being added to SK, which support function calling, introduce their specific models for function calling. The design, where each new connector introduces its own specific model class for function calling, does not scale well from the connector development perspective and does not allow for polymorphic use of connectors by the SK consumer code.

This ADR describes the high-level details of the service-agnostic function-calling model classes, while leaving the low-level details to the implementation phase. Additionally, this ADR outlines the identified options for various aspects of the design.

Requirements - https://github.com/microsoft/semantic-kernel/issues/5153

### CODE REVIEW

File: docs/decisions/0039-function-call-content.md [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, new ChatMessageContentItemCollection { new FunctionResultContent(functionCall, result) }));
+    
+    // of cause, the syntax can be simplified by having additional instance/extension methods
+    chatHistory.AddFunctionResultMessage(result); // Using the new AddFunctionResultMessage extension method of ChatHistory class
+}
+```
+
+Questions:
+- How to pass the original `FunctionCallContent` to connectors along with the function result. It's actually not clear atm whether it's needed or not. The current rationale is that some models might expect properties of the original function call, such as arguments, to be passed back to the LLM along with the function result. An argument can be made that the original function call can be found in the chat history by the connector if needed. However, a counterargument is that it may not always be possible because the chat history might be truncated to save tokens, reduce hallucination, etc.
+- How to pass function id to connector?
```
Comment (matthewbolanos) [inline docs/decisions/0039-function-call-content.md:348]: Couple of options from my side...
- Introduce a `public class FunctionCallResultContent : FunctionResult` that includes the ID (and perhaps the original `FunctionCallContent` too). This would be returned by the `functionCall.InvokeAsync` method.
- Introduce an optional `ID` property on `FunctionResult` (not the biggest fan of this...)

File: docs/decisions/0039-function-call-content.md [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+- How to communicate exception to the connectors? It was proposed to add the `Exception` property the the `FunctionResult` class that will always be assigned by the `KernelFunction.InvokeAsync` method. However, this change will break C# function calling semantic, where the function should be executed if the contract is satisfied, or an exception should be thrown if the contract is not fulfilled.
```
Comment (matthewbolanos) [inline docs/decisions/0039-function-call-content.md:349]: Based on the filter work that @dmytrostruk is driving, I _think_ we want to have the following principle...
- Always throw an error (to match standard C# behavior)
- If invoked within the context of a kernel (e.g., prompt template or function calling), have a filter that exposes that the error occurred.
- At this point, the developer should have two options:
  1. Do nothing, in this case, the error should bubble up so it can be caught elsewhere
  2. Handle the error so that the error reason can be passed back to the AI

If we did this, would we even _need_ the `Exception` property since the exception reason would just be put in the main result field?

I'm assuming the main challenge would be maintaining backwards compatibility where we automatically handle exceptions on behalf of the user. Let me think on this some more...

File: docs/decisions/0039-function-call-content.md [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+- If `FunctionResult` becomes a non-steaming content by inheriting `KernelContent` class, how the `FunctionResult` can represent streaming content capabilities represented by the `StreamingKernelContent` class when/if it needed later? C# does not support multiple inheritance.
```
Comment (matthewbolanos) [inline docs/decisions/0039-function-call-content.md:350]: This is a good one... Let me identify scenarios where this would be relevant.

File: docs/decisions/0039-function-call-content.md (hunk 1 above)
Comment (SergeyMenshykh) [reply docs/decisions/0039-function-call-content.md:348]: At the moment `FunctionResult` is not {de}serializable. Let me prototype to see what it would take to make it {de}serialization friendly.

File: docs/decisions/0039-function-call-content.md (hunk 2 above)
Comment (SergeyMenshykh) [reply docs/decisions/0039-function-call-content.md:349]: > If we did this, would we even _need_ the `Exception` property since the exception reason would just be put in the main result field?

if we do the `exception -> error message` translation inside a function filter and return the error message from the filter in a main result field, we won't need the `Exception` property.

> 
> I'm assuming the main challenge would be maintaining backwards compatibility where we automatically handle exceptions on behalf of the user. Let me think on this some more...

Do/will we handle exceptions automatically on behalf of users? Users will have to opt in by registering function filters to swallow/translate exceptions into error messages, won't they?

File: docs/decisions/0039-function-call-content.md (hunk 1 above)
Comment (SergeyMenshykh) [reply docs/decisions/0039-function-call-content.md:348]: Here are the results https://github.com/microsoft/semantic-kernel/pull/5703 of the exercise to make `FunctionResult` {de}serialization friendly:  
   
- The `FunctionResult.Function` property is lost (the property value is null) during the serialization/deserialization process, as it is not possible to serialize executable code. We can try to preserve the function parameters/return type schema, but it may require custom serialization, and it's not clear if it's worth it.  
   
- The `FunctionResult.Function` property had to be changed to nullable; otherwise, `FunctionResult` deserialization fails because the property was originally serialized as null and serializer can't create function instance from null. As a result, the change propagated to the `FunctionFilterContext.Function` property, which also became nullable. This may be considered a breaking change since filter code did not expect the `Function` property to be null, and suddenly after the SK update, it's null. However, I don't see it having a null value, since the filters are executed on a valid function before the function invocation returns a result that might get serialized if added to chat history, and only after deserialization will the `FunctionResult.Function` property become null.  
   
- The `FunctionResult.Value` loses its original type during the deserialization process. So, in [the unit test](https://github.com/microsoft/semantic-kernel/pull/5703/files#r1544994422), the Guid type turned into a JsonElement because the type of the value property is an object, and the best the JsonDeserializer can do in the absence of a more specific target type is to deserialize the JSON to `JsonElement`.  
   
In summary, yes, `FunctionResult` can be serializable/deserializable and, correspondingly, can become a derivative of the `KernelContent` class and can be used as a base class - `public class FunctionCallResultContent : FunctionResult`. 

However, the cost of the change would be the loss of the `Function` property (which might not be needed by connectors, and if needed it can be found via `kernel.Plugins` connectors have access to) and unnecessary propagation of the nullable nature of the `FunctionResult.Function` property to the `FunctionFilterContext.Function` property.

Next steps to prototype:
- Inherit the `FunctionResult` from `KernelContent`.
- Inherit the `Function{Call}ResultContent` from `FunctionResult`.

File: docs/decisions/0039-function-call-content.md [hunk 4]
```diff
@@ -0,0 +1,447 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
```
Comment (markwallace-microsoft) [inline docs/decisions/0039-function-call-content.md:3]: Update status and date

File: docs/decisions/0039-function-call-content.md [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+chatHistory.Add(new ChatMessageContent(AuthorRole.Tool, items);
+
+// Sending chat history containing function calls and function results to the LLM to get the final response
+messageContent = await completionService.GetChatMessageContentAsync(chatHistory, settings, kernel);
+```
+
+### Decision Outcome
+Option 1.2 was chosen due to its explicit nature.
+
+## 2. Function calling protocol for chat completion connectors
```
Comment (markwallace-microsoft) [inline docs/decisions/0039-function-call-content.md:208]: Should this be `2.`

File: docs/decisions/0039-function-call-content.md (hunk 4 above)
Comment (SergeyMenshykh) [reply docs/decisions/0039-function-call-content.md]: Done

**Review by dmytrostruk — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]
